scripts/scratch_ingest/expression_writer.py

```diff
@@ -0,0 +1,416 @@
+        )
+        pool.map(processor, slice_indexes)
+
+    def read_dense_matrix_slice(self, matrix_cells, cluster_cells, cluster_name, data_dir, indexes):
+        """
+        Read a dense matrix using start/stop indexes and create to individual gene-level files
+        :param matrix_cells: (List) cell names from matrix file
+        :param cluster_cells: (List) cell names from cluster file
+        :param cluster_name: (String) name of cluster object
+        :param data_dir: (String) name out output dir
+        :param indexes: (List) start/stop index points to read from/to
+        """
+        start_pos, end_pos = indexes
+        print(f"reading {self.matrix_file_path} at index {start_pos}:{end_pos}")
+        with ExpressionWriter.open_file(self.matrix_file_path) as matrix_file:
+            current_pos = start_pos
+            matrix_file.seek(current_pos)
+            while current_pos < end_pos:
+                line = matrix_file.readline()
+                self.process_dense_line(line, matrix_cells, cluster_cells, cluster_name, data_dir)
+                current_pos += len(line)
+
+    def process_dense_line(self, line, matrix_cells, cluster_cells, cluster_name, data_dir):
+        """
+        Process a single line from a dense matrix and write gene-level data
+
+        :param matrix_cells: (List) cells from header line of dense matrix
+        :param cluster_cells: (List) cell names from cluster file
+        :param cluster_name: (String) name of cluster object
+        :param data_dir (String) name out output dir
+        :param line: (String) single line from dense matrix
+        """
+        clean_line = line.rstrip().replace('"','')
+        raw_vals = re.split(COMMA_OR_TAB, clean_line)
+        gene_name = raw_vals.pop(0)
+        exp_vals = [round(float(val), self.precision) if float(val) != 0.0 else 0 for val in raw_vals]
+        filtered_expression = ExpressionWriter.filter_expression_for_cluster(
+            cluster_cells, matrix_cells, exp_vals
+        )
+        ExpressionWriter.write_gene_scores(cluster_name, gene_name, filtered_expression, data_dir)
+
+    @staticmethod
+    def filter_expression_for_cluster(cluster_cells, exp_cells, exp_scores):
+        """
+        Assemble a List of expression scores, filtered & ordered from a List of cluster cells
+        Will substitute 0 as a value for any cell not seen in the expression file
+
+        :param cluster_cells: (List) cluster cell names
+        :param exp_cells: (List) expression cell names
+        :param exp_scores: (List) expression values, in the same order as exp_cells
+        :return: (List) Expression values, ordered by cluster_cells
+        """
+        observed_exp = dict(zip(exp_cells, exp_scores))
+        return (observed_exp.get(cell, 0) for cell in cluster_cells)
+
+    @staticmethod
+    def write_gene_scores(cluster_name, gene_name, exp_values, data_dir):
```

Comment:
Optimization idea: upload small batches of files sooner instead of all files last.  This method seems relevant there.

To start, I imagine we'd write all expression array files to disk, then upload (`gsutil -m cp...`) them all to the bucket.  That approach would make any per-gene downstream pipelines (e.g. Image Pipeline) wait until the scatter data pipeline totally completes.  To allow downstream pipelines to start sooner, this pipeline could -- in a background thread -- upload expression array files in sets of, say, 100, to the bucket, while updating a file listing uploaded genes.  (This `write_gene_scores` method would presumably increment a related counter.)  That would let downstream pipelines pull the gene list index, then process expression arrays as they become available.

This optimization would decrease image cache warming times likely by > 1 hour for this pilot case.

If that sounds feasible to you, I'd be happy to create a ticket, then we can further assess effort-value tradeoffs and priority relative to other optimizations.

Reply:
I'm a little less enthused about this one, though on its face it makes sense.  I'm just worried about concurrency/polling.  It is much easier to architect as successive jobs being dispatched from the portal, instead of two concurrent jobs.  That being said it's definitely still worth ticketing and we can prioritize as we see fit.